### PR TITLE
Merge various quality-of-life changes & tides boundary functions (From GFDL NWA25) into RMOM6

### PR DESCRIPTION
**Pull request by manishvenu**
This (originally draft) PR is a conglomeration of work by @ashjbarnes and me adding tidal boundary conditions, certain MOM file functions, and other quality of life improvements to RM6. Here is a categorization of the changes and TO-DOs:

1. Add tides from GFDL NWA25 to RMOM6 in 5 functions: 
    - [X] Two util functions  (utils.ap2ep, utils.ep2ap)
    - [X] Coordinate function for the boundaries (segment.coords)
    - [X] Implement loose horizontal subsetting (utils.find_roughly_nearest_ny_nx)
    - [X] Create a setup tides function for the user side (expt.setup_tides_rectangular_boundaries)
    - [X] Create a regrid tides function to setup the tides on the RMOM6 generated Hgrid (segment.regrid_tides)
     - [ ] Use segment.coords for the expt.ocean_state_rectangular_boundaries (changed to a new branch w/ new PR #13 )
     - [ ] Use segment.coords to read-in angled grids (like NWA12) (changed to a new branch w/ new PR #13 )
     - [x]  Test tides in the CESM
     - [x] Add test tides functions to rm6 tests

2. Change function names to verb-starts & more descriptive names

    - [X] Change expt.rectangular_boundaries and expt.simple_boundaries to "setup_ocean_state_...."
    - [ ] Create a coherent and understandable naming scheme, which will address #9, shifted to PR#13
    - [ ] Validate function name choices

3. Make the changes to MOM files caused by RMOM6 more understandable (addresses #8)

    - [X] Create MOM file dictionary for better manipulation of MOM_input, override, and layout
    - [x] Rework the setup_run_directory to more clearly make changes in MOM_override vs changing MOM_input
    - [x] Create config files, or json/another abstraction layer to explain the changes caused by RMOM6
    - [x] As part of the previous, create a feature/function for the user to change parameters (See change_MOM_parameter)
    - [ ] Verify this fix addresses #8 

4. Extraneous small changes

    - [x] Remove greek letters

### CODE REVIEW

**Comment by manishvenu** [comment]
Hey @ashjbarnes, 

Let me know if you have any opinions on what should be added or taken out from here!

**Comment by manishvenu** [comment]
I don't know want to try to remove the OBC stuff from the default MOM_input files b/c not sure if I 'll miss a place they are stored based on if installed as a package or cloned, so I'll manually delete those fields in setup_run_directory.

**Comment by ashjbarnes** [comment]
> I don't want to try to remove the OBC stuff from the default MOM_input files b/c not sure if I 'll miss a place they are stored based on if installed as a package or cloned, so I'll manually delete those fields in setup_run_directory.

Should be fine to delete them - the only difference with installed vs not installed is to do with the directory structure of the package. Either way, it'll always be in the `premade_run_directories` folder. So you can delete if you'd prefer! That's probably cleaner than doing programatically each time

File: regional_mom6/regional_mom6.py [hunk 1]
```diff
@@ -14,20 +14,201 @@
...
+from collections import defaultdict
+import json
 
 warnings.filterwarnings("ignore")
 
 __all__ = [
     "longitude_slicer",
     "hyperbolictan_thickness_profile",
-    "rectangular_hgrid",
+    "calculate_rectangular_hgrid",
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:36]:
```suggestion
    "generate_rectangular_hgrid",
```

Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:36]:
```suggestion
    "generate_rectangular_hgrid",
```

File: regional_mom6/regional_mom6.py [hunk 2]
```diff
@@ -170,17 +348,17 @@ def get_glorys_data(
     """
     buffer = 0.24  # Pads downloads to ensure that interpolation onto desired domain doesn't fail. Default of 0.24 is twice Glorys cell width (12th degree)
 
-    path = Path(download_path)
+    path = os.path.join(download_path)
 
     if modify_existing:
-        file = open(path / "get_glorysdata.sh", "r")
+        file = open(os.path.join(path, "get_glorysdata.sh"), "r")
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:354]: Likewise

**Comment by manishvenu** [comment]
> > I don't want to try to remove the OBC stuff from the default MOM_input files b/c not sure if I 'll miss a place they are stored based on if installed as a package or cloned, so I'll manually delete those fields in setup_run_directory.
> 
> Should be fine to delete them - the only difference with installed vs not installed is to do with the directory structure of the package. Either way, it'll always be in the `premade_run_directories` folder. So you can delete if you'd prefer! That's probably cleaner than doing programatically each time

Sounds good, I'll take it out of the premade_run_directories folder, but keep in the deletion to explicitly state it in setup_run_directories.

File: regional_mom6/regional_mom6.py [hunk 3]
```diff
@@ -170,17 +348,17 @@ def get_glorys_data(
...
         lines = file.readlines()
         file.close()
 
     else:
         lines = ["#!/bin/bash\n"]
 
-    file = open(path / "get_glorysdata.sh", "w")
+    file = open(os.path.join(path, "get_glorysdata.sh"), "w")
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:361]: Likewise

File: regional_mom6/regional_mom6.py [hunk 4]
```diff
@@ -308,10 +486,10 @@ def hyperbolictan_thickness_profile(nlayers, ratio, total_depth):
     return layer_thicknesses
 
 
-def rectangular_hgrid(λ, φ):
+def calculate_rectangular_hgrid(lons, lats):
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:489]:
```suggestion
def generate_rectangular_hgrid(lons, lats):
```

File: regional_mom6/regional_mom6.py [hunk 5]
```diff
@@ -584,11 +851,11 @@ def _make_hgrid(self):
             if ny % 2 != 1:
                 ny += 1
 
-            φ = np.linspace(
+            lats = np.linspace(
                 self.latitude_extent[0], self.latitude_extent[1], ny
             )  # latitudes in degrees
 
-            hgrid = rectangular_hgrid(λ, φ)
+            hgrid = calculate_rectangular_hgrid(lons, lats)
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:858]:
```suggestion
            hgrid = generate_rectangular_hgrid(lons, lats)
```

File: regional_mom6/regional_mom6.py [hunk 6]
```diff
@@ -170,17 +348,17 @@ def get_glorys_data(
     """
     buffer = 0.24  # Pads downloads to ensure that interpolation onto desired domain doesn't fail. Default of 0.24 is twice Glorys cell width (12th degree)
 
-    path = Path(download_path)
+    path = os.path.join(download_path)
```
[the remark this answers is not in the document]
Comment (manishvenu) [reply regional_mom6/regional_mom6.py:351]: Shoot, so sorry, that's totally on me

File: regional_mom6/regional_mom6.py [hunk 7]
```diff
@@ -619,13 +886,187 @@ def _make_vgrid(self):
...
+            # return datasets
+
+            return all_files
+        except:
+            return "No initial condition set up yet (or files misplaced from {}). Call `setup_initial_condition` method to set up initial conditions.".format(
+                self.mom_input_dir / "forcing"
+            )
+
+    @property
+    def bathymetry_property(self):
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:986]: Just wondering if we need the _property suffix? Or would calling `expt.bathymetry` be a bit more intuitive to match the others like `expt.initial_condition`?

File: regional_mom6/regional_mom6.py [hunk 8]
```diff
@@ -1099,10 +1561,134 @@ def simple_boundary(
...
+            .rename({"lon_z": "lon", "lat_z": "lat", "nc": "constituent"})
+            .isel(
+                constituent=convert_to_tpxo_tidal_constituents(self.tidal_constituents)
+            )
+        )
+        tidal_360_lon = [
+            self.longitude_extent[0],
+            self.longitude_extent[1],
+        ]
+        ny0, nx0 = find_roughly_nearest_ny_nx(
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:1621]: This line crashed when testing with the NWA run. Perhaps an issue with the curved boundary? I'm looking into it

File: regional_mom6/regional_mom6.py [hunk 9]
```diff
@@ -1580,7 +2167,11 @@ def setup_run_directory(
 
         ## Get the path to the regional_mom package on this computer
         premade_rundir_path = Path(
-            importlib.resources.files("regional_mom6") / "demos/premade_run_directories"
+            os.path.join(
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:2169]: Would prefer using Path

File: regional_mom6/regional_mom6.py [hunk 10]
```diff
@@ -1589,24 +2180,31 @@ def setup_run_directory(
             )
 
             premade_rundir_path = Path(
-                importlib.resources.files("regional_mom6").parent
-                / "demos/premade_run_directories"
+                os.path.join(
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:2182]: likewise

File: regional_mom6/regional_mom6.py [hunk 11]
```diff
@@ -1589,24 +2180,31 @@ def setup_run_directory(
...
+                    "premade_run_directories",
+                )
             )
             if not premade_rundir_path.exists():
                 raise ValueError(
                     f"Cannot find the premade run directory files at {premade_rundir_path} either.\n\n"
                     + "There may be an issue with package installation. Check that the `premade_run_directory` folder is present in one of these two locations"
                 )
+            else:
+                print("Found run files. Continuing...")
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:2194]: nice this was also really annoying for me!!

File: regional_mom6/regional_mom6.py [hunk 12]
```diff
@@ -1589,24 +2180,31 @@ def setup_run_directory(
...
 
         # Define the locations of the directories we'll copy files across from. Base contains most of the files, and overwrite replaces files in the base directory.
-        base_run_dir = premade_rundir_path / "common_files"
+        base_run_dir = Path(os.path.join(premade_rundir_path, "common_files"))
```
Comment (ashjbarnes) [inline regional_mom6/regional_mom6.py:2197]: Path object

File: tests/test_grid_generation.py [hunk 13]
```diff
@@ -2,7 +2,7 @@
 import pytest
 
 from regional_mom6 import hyperbolictan_thickness_profile
-from regional_mom6 import rectangular_hgrid
+from regional_mom6 import calculate_rectangular_hgrid
```
Comment (ashjbarnes) [inline tests/test_grid_generation.py:5]:
```suggestion
from regional_mom6 import generate_rectangular_hgrid
```

**Review by ashjbarnes — CHANGES_REQUESTED** [review]
This is awesome Manish!! Aside from a few comments I've left, main thing I see before reviewing is:

 - Test the longitude slicer on a domain that cuts across the 'seam' of TPXO to make sure this won't be an issue
  - Test the 'read from existing config file' stuff on an existing run to double check the workflow accurately clones the model. You may have already done this

File: regional_mom6/regional_mom6.py (hunk 8 above)
Comment (manishvenu) [reply regional_mom6/regional_mom6.py:1621]: Deleted b/c not necessarily needed, added to next PR

**Review by ashjbarnes — APPROVED** [review]

**Pull request merged**